<a href="https://colab.research.google.com/github/ladiaJeong/-01-classification/blob/master/%5B01%5Dclassification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 1) load_digits : 손글씨를 분류해 봅시다

In [ ]:
import sklearn

print(sklearn.__version__)


1.0


### (1) 필요한 모듈 import하기

In [ ]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

### (2) 데이터 준비

In [ ]:
digits = load_digits()
print(dir(digits))#Feature Data 지정하기

['DESCR', 'data', 'feature_names', 'frame', 'images', 'target', 'target_names']


### (3) 데이터 이해하기


In [ ]:
digits_data = digits.data
print(digits_data.shape)

(1797, 64)


In [ ]:
digits_label = digits.target #Label Data 지정하기

In [ ]:
digits.target_names #Target Names 출력해 보기

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
print(digits.DESCR) #데이터 Describe 해 보기

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

### (4) train, test 데이터 분리

In [ ]:
import pandas as pd

print(pd.__version__)

1.3.3


In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(digits_data,  #feature
                                                    digits_label,  #label
                                                    test_size=0.2,   #test dataset 크기
                                                    random_state=7)  #데이터 분리 시 랜덤성

print('X_train 개수: ', len(X_train),', X_test 개수: ', len(X_test))                                                   

X_train 개수:  1437 , X_test 개수:  360


### (5) 다양한 모델로 학습시켜보기

In [ ]:
#Decision Tree 사용해 보기
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.81      0.81      0.81        42
           2       0.79      0.82      0.80        40
           3       0.79      0.91      0.85        34
           4       0.83      0.95      0.89        37
           5       0.90      0.96      0.93        28
           6       0.84      0.93      0.88        28
           7       0.96      0.82      0.89        33
           8       0.88      0.65      0.75        43
           9       0.78      0.78      0.78        32

    accuracy                           0.86       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.86      0.86      0.85       360

0.8555555555555555


In [ ]:
#Random Forest 사용해 보기
from sklearn.ensemble import RandomForestClassifier #랜덤포레스트라는 분류기를 사용하기 위해 import

X_train, X_test, y_train, y_test = train_test_split(digits_data, # iris 데이터의 data 컬럼
                                                    digits_label, # iris 데이터의 target 컬럼
                                                    test_size=0.2, # test_size : train data와 test data를 몇대몇으로 나눌지 정하는 옵션
                                                    random_state=21) # random_state : 랜덤 패턴의 값을 지정

random_forest = RandomForestClassifier(random_state=32) # RandomForest분류기 객체를 생성
random_forest.fit(X_train, y_train) # 훈련
y_pred = random_forest.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       0.97      0.94      0.95        32
           1       0.97      1.00      0.99        36
           2       1.00      1.00      1.00        30
           3       0.98      0.98      0.98        41
           4       0.94      0.97      0.95        32
           5       1.00      0.98      0.99        46
           6       1.00      0.97      0.98        32
           7       0.98      1.00      0.99        40
           8       0.93      0.98      0.95        42
           9       1.00      0.93      0.96        29

    accuracy                           0.97       360
   macro avg       0.98      0.97      0.97       360
weighted avg       0.98      0.97      0.98       360

0.975


In [ ]:
#SVM 사용해 보기
from sklearn import svm #Support Vector Machine을 사용하기 위해 import
svm_model = svm.SVC() # 모델 객체를 만든다.

print(svm_model._estimator_type) # 이 모델이 어떤 타입인지 확인한다.

svm_model.fit(X_train, y_train) # 훈련
y_pred = svm_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred)) #정확도 확인

classifier
              precision    recall  f1-score   support

           0       1.00      0.97      0.98        32
           1       0.95      1.00      0.97        36
           2       1.00      1.00      1.00        30
           3       1.00      1.00      1.00        41
           4       0.97      1.00      0.98        32
           5       0.98      1.00      0.99        46
           6       1.00      1.00      1.00        32
           7       1.00      1.00      1.00        40
           8       0.98      0.95      0.96        42
           9       1.00      0.93      0.96        29

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360

0.9861111111111112


In [ ]:
#SGD Classifier 사용해 보기
from sklearn.linear_model import SGDClassifier #선형분류기인 SGDClassifier를 사용하기 위한 import
sgd_model = SGDClassifier() # 모델 객체 생성

print(sgd_model._estimator_type) # 이 모델의 타입을 확인

sgd_model.fit(X_train, y_train) # sgd모델로 훈련데이터로 훈련시킨다.
y_pred = sgd_model.predict(X_test)# 그 모델로 test데이터를 사용해 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred)) #정확도 확인

classifier
              precision    recall  f1-score   support

           0       1.00      0.94      0.97        32
           1       0.92      0.92      0.92        36
           2       1.00      1.00      1.00        30
           3       0.93      1.00      0.96        41
           4       0.91      1.00      0.96        32
           5       0.98      0.98      0.98        46
           6       0.94      1.00      0.97        32
           7       1.00      0.97      0.99        40
           8       1.00      0.88      0.94        42
           9       0.97      0.97      0.97        29

    accuracy                           0.96       360
   macro avg       0.96      0.97      0.96       360
weighted avg       0.97      0.96      0.96       360

0.9638888888888889


In [ ]:
#Logistic Regression 사용해 보기
from sklearn.linear_model import LogisticRegression # 선형분류기인 LogisticRegression를 사용하기 위한 import
logistic_model = LogisticRegression() # 모델 객체 생성

print(logistic_model._estimator_type) # 이 모델의 타입을 확인

logistic_model.fit(X_train, y_train) #LogisticRegression모델로 훈련데이터를 가지고 훈련시킨다.
y_pred = logistic_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred)) #정확도 확인

classifier
              precision    recall  f1-score   support

           0       0.92      0.85      0.88        39
           1       0.92      0.96      0.94        75

    accuracy                           0.92       114
   macro avg       0.92      0.90      0.91       114
weighted avg       0.92      0.92      0.92       114

0.9210526315789473


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


### (6) 모델을 평가해 보기

손글씨 데이터는 총 1,797개, 62개의 숫자로 이루어져있다. 각 이미지가 62(8x8)개의 픽셀을 일렬로 펴놓은것이다. target 데이터는 0부터 9까지의 숫자를 나타낸다.

Decision tree :0.844  
random forest :0.975  
svm :0.986  
sgd classifier :0.958  
Logistic Regression :0.978 (소수 3째자리까지 반올림으로 나타내기)

## 회고

### 1. 손글씨 데이터에서 모델 성능 평가 수치를 보면 눈에 띄는 것이 의사결정 트리 모델 성능이 가장 낮은 것이다. 
손글씨 데이터는 특정 숫자가 T or F 의 값 중 하나를 선택해서 결과값을 도출하는 방식이다. 마치 토너먼트 경기처럼 한번 지면 다시 기회가 없듯 한번 잘못 판단하는 데이터는 다시 기회가 없기 때문에 손실이 발생하는 것 같다.  이를 보완한 랜덤 포레스트는 데이터에게 랜덤으로 기회를 여러번 주기때문에 손실이 적게 발생하고 모델 성능이 의사결정 나무에 비해 훨씬 좋다.  

랜덤 포레스트 방식을 통해 '컴퓨터적 사고'에 대해 생각해 보았다. 의사결정 트리의 단점 보완을 위해서 다른 방법을 찾기 보다는 같은 로직 안에서 [랜덤+무한반복]이라는 방식의 도구로 해결한 것이 참 좋았다. 숲에서 나오는 여러 의견을을 통합하고 합쳐서 결과값을 내니 모델의 성능이 좋아진다.  

### 2. Logistic Regression 를 사용했을 때 핑크색 박스에 알 수 없는 정보가 가득했다. 

일단은 와인과 유방암에서도 나오는 지 비교하기 위해 그냥 두었다. 

### 3. 가장 성능이 좋았던 모델은 svm이다. 

---

## 2) load_wine : 와인을 분류해 봅시다

### (1) 필요한 모듈 import하기

In [ ]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

### (2) 데이터 준비

In [ ]:
wine = load_wine()
print(dir(wine))#Feature Data 지정하기

wine_data = wine.data

print(wine_data.shape)


['DESCR', 'data', 'feature_names', 'frame', 'target', 'target_names']
(178, 13)


### (3) 데이터 이해하기


In [ ]:
#Feature Data 지정하기
wine_label = wine.target #Label Data 지정하기
wine.target_names #Target Names 출력해 보기

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [ ]:
print(wine.DESCR) #데이터 Describe 해 보기

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

### (4) train, test 데이터 분리

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(wine_data,  #feature
                                                    wine_label,  #label
                                                    test_size=0.2,   #test dataset 크기
                                                    random_state=7)  #데이터 분리 시 랜덤성

print('X_train 개수: ', len(X_train),', X_test 개수: ', len(X_test))                                                   

X_train 개수:  142 , X_test 개수:  36


### (5) 다양한 모델로 학습시켜보기

In [ ]:
#Decision Tree 사용해 보기
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       0.90      0.92      0.91        39
           1       0.96      0.95      0.95        75

    accuracy                           0.94       114
   macro avg       0.93      0.93      0.93       114
weighted avg       0.94      0.94      0.94       114

0.9385964912280702


In [ ]:
#Random Forest 사용해 보기
from sklearn.ensemble import RandomForestClassifier #랜덤포레스트라는 분류기를 사용하기 위해 import

X_train, X_test, y_train, y_test = train_test_split(wine_data, # iris 데이터의 data 컬럼
                                                    wine_label, # iris 데이터의 target 컬럼
                                                    test_size=0.2, # test_size : train data와 test data를 몇대몇으로 나눌지 정하는 옵션
                                                    random_state=7) # random_state : 랜덤 패턴의 값을 지정

random_forest = RandomForestClassifier(random_state=32) # RandomForest분류기 객체를 생성
random_forest.fit(X_train, y_train) # 훈련
y_pred = random_forest.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       1.00      1.00      1.00        17
           2       1.00      1.00      1.00        12

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36

1.0


In [ ]:
#SVM 사용해 보기
from sklearn import svm #Support Vector Machine을 사용하기 위해 import
svm_model = svm.SVC() # 모델 객체를 만든다.

print(svm_model._estimator_type) # 이 모델이 어떤 타입인지 확인한다.

svm_model.fit(X_train, y_train) # 훈련
y_pred = svm_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred)) #정확도 확인

classifier
              precision    recall  f1-score   support

           0       0.86      0.86      0.86         7
           1       0.58      0.88      0.70        17
           2       0.33      0.08      0.13        12

    accuracy                           0.61        36
   macro avg       0.59      0.61      0.56        36
weighted avg       0.55      0.61      0.54        36

0.6111111111111112


In [ ]:
#SGD Classifier 사용해 보기
from sklearn.linear_model import SGDClassifier #선형분류기인 SGDClassifier를 사용하기 위한 import
sgd_model = SGDClassifier() # 모델 객체 생성

print(sgd_model._estimator_type) # 이 모델의 타입을 확인

sgd_model.fit(X_train, y_train) # sgd모델로 훈련데이터로 훈련시킨다.
y_pred = sgd_model.predict(X_test)# 그 모델로 test데이터를 사용해 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred)) #정확도 확인

classifier
              precision    recall  f1-score   support

           0       0.92      0.73      0.81        15
           1       0.42      1.00      0.59        10
           2       0.00      0.00      0.00        11

    accuracy                           0.58        36
   macro avg       0.44      0.58      0.47        36
weighted avg       0.50      0.58      0.50        36

0.5833333333333334


/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
#Logistic Regression 사용해 보기
from sklearn.linear_model import LogisticRegression # 선형분류기인 LogisticRegression를 사용하기 위한 import
logistic_model = LogisticRegression(max_iter = 4000) # 모델 객체 생성

print(logistic_model._estimator_type) # 이 모델의 타입을 확인

logistic_model.fit(X_train, y_train) #LogisticRegression모델로 훈련데이터를 가지고 훈련시킨다.
y_pred = logistic_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred)) #정확도 확인

classifier
              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       0.94      1.00      0.97        17
           2       1.00      0.92      0.96        12

    accuracy                           0.97        36
   macro avg       0.98      0.97      0.98        36
weighted avg       0.97      0.97      0.97        36

0.9722222222222222


### (6) 모델을 평가해 보기

총 178개의 데이터 feature는 13개로 Alcohol, Malic acid, Color intensity 등 와인의 특성값과 label은 class 0,1,2로 세개가 있다.  

Decision tree :0.944  
random forest :1.000  
svm :0.611  
sgd classifier :0.583  
Logistic Regression :0.972 (소수 3째자리까지 반올림으로 나타내기)

## 회고

### 1. 와인 데이터에서 주목할 만한 것은 svm과 sgd의 모델의 성능이 낮다는 것이다. 와인의 등급을 판정하는 것은 오차행렬을 적합하기에 스팸메일이나 유방암 판정같은 상황은 아니다. 

### 2.가장 성능이 좋은 모델은 랜덤 포레스트이다. 분류하는 데이터여서 의사결정 나무 방식이 성능이 좋은 것 같다. 

---

## 3) load_breast_cancer : 유방암 여부를 진단해 봅시다

### (1) 필요한 모듈 import하기

In [ ]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

### (2) 데이터 준비

In [ ]:
breast_cancer = load_breast_cancer()
print(dir(breast_cancer))#Feature Data 지정하기

breast_cancer_data = breast_cancer.data

print(breast_cancer_data.shape)


['DESCR', 'data', 'data_module', 'feature_names', 'filename', 'frame', 'target', 'target_names']
(569, 30)


### (3) 데이터 이해하기


In [ ]:
#Feature Data 지정하기
breast_cancer_label = breast_cancer.target #Label Data 지정하기
breast_cancer.target_names #Target Names 출력해 보기

array(['malignant', 'benign'], dtype='<U9')

In [ ]:
#데이터 Describe 해 보기
print(breast_cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

### (4) train, test 데이터 분리

In [ ]:
#trian, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data,  #feature
                                                    breast_cancer_label,  #label
                                                    test_size=0.2,   #test dataset 크기
                                                    random_state=7)  #데이터 분리 시 랜덤성


print('X_train 개수: ', len(X_train),', X_test 개수: ', len(X_test))      

X_train 개수:  455 , X_test 개수:  114


### (5) 다양한 모델로 학습시켜보기

In [ ]:
#Decision Tree 사용해 보기
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       0.92      0.82      0.87        40
           1       0.91      0.96      0.93        74

    accuracy                           0.91       114
   macro avg       0.91      0.89      0.90       114
weighted avg       0.91      0.91      0.91       114

0.9122807017543859


In [ ]:
#Random Forest 사용해 보기
from sklearn.ensemble import RandomForestClassifier #랜덤포레스트라는 분류기를 사용하기 위해 import

X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data, # iris 데이터의 data 컬럼
                                                    breast_cancer_label, # iris 데이터의 target 컬럼
                                                    test_size=0.2, # test_size : train data와 test data를 몇대몇으로 나눌지 정하는 옵션
                                                    random_state=21) # random_state : 랜덤 패턴의 값을 지정

random_forest = RandomForestClassifier(random_state=32) # RandomForest분류기 객체를 생성
random_forest.fit(X_train, y_train) # 훈련
y_pred = random_forest.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       0.97      0.95      0.96        39
           1       0.97      0.99      0.98        75

    accuracy                           0.97       114
   macro avg       0.97      0.97      0.97       114
weighted avg       0.97      0.97      0.97       114

0.9736842105263158


In [ ]:
#SVM 사용해 보기
from sklearn import svm #Support Vector Machine을 사용하기 위해 import
svm_model = svm.SVC() # 모델 객체를 만든다.

print(svm_model._estimator_type) # 이 모델이 어떤 타입인지 확인한다.

svm_model.fit(X_train, y_train) # 훈련
y_pred = svm_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred)) #정확도 확인

classifier
              precision    recall  f1-score   support

           0       0.93      0.72      0.81        39
           1       0.87      0.97      0.92        75

    accuracy                           0.89       114
   macro avg       0.90      0.85      0.86       114
weighted avg       0.89      0.89      0.88       114

0.8859649122807017


In [ ]:
#SGD Classifier 사용해 보기
from sklearn.linear_model import SGDClassifier #선형분류기인 SGDClassifier를 사용하기 위한 import
sgd_model = SGDClassifier() # 모델 객체 생성

print(sgd_model._estimator_type) # 이 모델의 타입을 확인

sgd_model.fit(X_train, y_train) # sgd모델로 훈련데이터로 훈련시킨다.
y_pred = sgd_model.predict(X_test)# 그 모델로 test데이터를 사용해 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred)) #정확도 확인

classifier
              precision    recall  f1-score   support

           0       0.89      0.79      0.84        39
           1       0.90      0.95      0.92        75

    accuracy                           0.89       114
   macro avg       0.89      0.87      0.88       114
weighted avg       0.89      0.89      0.89       114

0.8947368421052632


In [ ]:
#Logistic Regression 사용해 보기
from sklearn.linear_model import LogisticRegression # 선형분류기인 LogisticRegression를 사용하기 위한 import
logistic_model = LogisticRegression(max_iter = 4000) # 모델 객체 생성

print(logistic_model._estimator_type) # 이 모델의 타입을 확인

logistic_model.fit(X_train, y_train) #LogisticRegression모델로 훈련데이터를 가지고 훈련시킨다.
y_pred = logistic_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred)) #정확도 확인

classifier
              precision    recall  f1-score   support

           0       0.92      0.87      0.89        39
           1       0.94      0.96      0.95        75

    accuracy                           0.93       114
   macro avg       0.93      0.92      0.92       114
weighted avg       0.93      0.93      0.93       114

0.9298245614035088


### (6) 모델을 평가해 보기


Decision tree :0.912  
random forest :0.974  
svm :0.886  
sgd classifier :0.895  
Logistic Regression :0.921 (소수 3째자리까지 반올림으로 나타내기)

## 회고

### 1. 핑크색 상자에 대한 부분은 검색을 통해 해결해 보었다. 디폴트 값 대신 (max_iter = 4000)를 넣어서 많이 학습을 시키는 방법으로 문제를 해결해 보았다. 

### 2.  random_state 값에 따라 결과값이 변화가 있을 듯 하다는 가설로 테스트를 해 보았다. 영향을 받는것 같다. 

### 3. 시작 전에는 무지 막막했지만, 코드를 다 이해하지는 못해도 반복하다보니 어렴풋이 알게 되는 것들이 있었다. 

### 4. 유방암 데이터는 recall 모델로 평가해야 한다. 양성인데 음성을 판단하면 안되기 때문이다. 

### 5. 뭐라도 짜내서 더 적고 싶지만 지금은 월요일 밤 11시 33분 이기에 마무리 한다. 데이터는 들여다 보면 볼 수록 많은 정보를 캐낼 수 있는 것 같다. 

#### 레퍼런스

##### https://github.com/yenaryu/AI/blob/main/%5BE-02%5Dload_digits%2Cwine%2Cbreast_cancer.ipynb 

https://velog.io/@ja2y5ung/E2.-sklearn%EC%9D%98-Toy-Data%EB%A5%BC-%EC%97%AC%EB%9F%AC-%EC%9D%B8%EC%8B%9D-%EB%AA%A8%EB%8D%B8%EC%97%90-%EB%84%A3%EC%96%B4%EB%B3%B4%EA%B8%B0 

https://velog.io/@oooops/%EB%A8%B8%EC%8B%A0%EB%9F%AC%EB%8B%9D%EC%9C%BC%EB%A1%9C-%EB%8D%B0%EC%9D%B4%ED%84%B0-%EB%B6%84%EB%A5%98%ED%95%B4%EB%B3%B4%EA%B8%B0
